### PR TITLE
MosaicML StreamingDataset (Part 2, vision datasets)

### PR DESCRIPTION
**Pull request by abhi-mosaic**
This PR adds implementations of Vision datasets using the new `StreamingDataset` class.

Specifically we will add streaming versions of ADE20k, ImageNet1k, and COCO, with more to come later.

Dataset writing scripts are kept out of the Composer core, as they are meant to be used when you have a local copy of the dataset available and are converting it to the `StreamingDataset` format.

### CODE REVIEW

File: composer/datasets/coco.py [hunk 1]
```diff
@@ -177,3 +181,102 @@ def split_dict_fn(batch: Batch, num_microbatches: int) -> Sequence[Batch]:  #typ
...
+        local (str): Local filesystem directory where dataset is cached during operation.
+    """
+
+    remote: str = hp.optional('Remote directory (S3 or local filesystem) where dataset is stored',
+                              default='s3://mosaicml-internal-dataset-coco/mds/')
+    local: str = hp.optional('Local filesystem directory where dataset is cached during operation',
+                             default='/tmp/mds-cache/mds-coco/')
+
+    def initialize_object(self, batch_size: int, dataloader_hparams: DataLoaderHparams):
+        dboxes = dboxes300_coco()
```
Comment (hanlint) [inline composer/datasets/coco.py:267]: same comment -> move this to StreamingCOCO

File: composer/datasets/ade20k.py [hunk 2]
```diff
@@ -388,6 +391,136 @@ def initialize_object(self, batch_size, dataloader_hparams) -> DataSpec:
...
+            raise ValueError("base_size cannot be zero or negative.")
+
+        if self.min_resize_scale <= 0:
+            raise ValueError("min_resize_scale cannot be zero or negative")
+
+        if self.max_resize_scale < self.min_resize_scale:
+            raise ValueError("max_resize_scale cannot be less than min_resize_scale")
+
+    def initialize_object(self, batch_size: int, dataloader_hparams: DataLoaderHparams) -> DataSpec:
+        self.validate()
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/datasets/ade20k.py:478]: Yes can do... I have a question though, do we want this preprocessing functionality like `split`, `base_size`, `min_resize_scale`, ... to all be part of our `StreamingADE20k` dataset? I'm in favor of this, just want to emphasize that now it won't just be an ADE20k dataset that returns PIL images, it's a dataset with some built in user-chosen preprocessing steps. So less reusable code, but more clearly an _example_ of how a user constructs their custom dataset.

[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/datasets/ade20k.py:478]: Also, dataloader construction will still live in `initialize_object` by necessity, because `dataloader_hparams` and `batch_size` are passed in via the Trainer

File: composer/datasets/ade20k.py [hunk 3]
```diff
@@ -388,6 +391,155 @@ def initialize_object(self, batch_size, dataloader_hparams) -> DataSpec:
...
+    Args:
+        remote (str): Remote directory (S3 or local filesystem) where dataset is stored.
+        local (str): Local filesystem directory where dataset is cached during operation.
+        split (str): The dataset split to use, either 'train', or 'val'. Default: ``'train```.
+        shuffle (bool): Whether to shuffle the samples in this dataset.
+        base_size (int): initial size of the image and target before other augmentations. Default: ``512``.
+        min_resize_scale (float): the minimum value the samples can be rescaled. Default: ``0.5``.
+        max_resize_scale (float): the maximum value the samples can be rescaled. Default: ``2.0``.
+        final_size (int): the final size of the image and target. Default: ``512``.
+    """
```
Comment (abhi-mosaic) [inline composer/datasets/ade20k.py:407]: Refactored so that the `StreamingADE20k` dataset does most of the work, making it easier to use in a standalone way.

Collating and device normalization are still done in the hparams `initialize_object` when we create the dataloader.

@hanlint Can I avoid refactoring the older `ImageFolder`-based ADE20k dataset in this PR? I an open a JIRA issue to track. Would prefer to make this PR simple and additive, and do edits/deletions in a separate one.

File: scripts/mds/ade20k.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    Args:
+        args (Namespace): Commandline arguments.
+    """
+    fields = ['uid', 'image', 'annotation']
+
+    for (split, expected_num_samples, shuffle) in [
+        ("train", 20206, True),
+        ("val", 2000, False),
+    ]:
```
Comment (abhi-mosaic) [inline scripts/mds/ade20k.py:83]: Only uploaded the `train` and `val` splits at the moment, as the `test` split doesn't have labels and I'm not sure how we would use it... if @Landanjs has a use case I can add support, there will be some extra logic on the Dataset init to make sure we don't try to decode keys that don't exist.

File: tests/datasets/test_streaming_remote.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+import pytest
+from PIL.Image import Image
+from torch.utils.data import DataLoader
+
+from composer.datasets.ade20k import StreamingADE20k
+from composer.datasets.utils import pil_image_collate
+
+
+# @pytest.mark.skip()
```
Comment (abhi-mosaic) [inline tests/datasets/test_streaming_remote.py:12]: @ravi-mosaicml could you provide me a quick example of how to add that @pytest.mark.integration thing? From what I understood I should add **both** the integration flag and the skip(), with the latter being removed once we have later updated our Jenkins workflow to recognize the integration.

File: tests/datasets/test_streaming_remote.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    for epoch in range(3):
+        rcvd_samples = 0
+        epoch_start = time.time()
+        for _, (images, _) in enumerate(loader):
+            n_samples = images.shape[0]
+            rcvd_samples += n_samples
+        epoch_end = time.time()
+        epoch_dur = epoch_end - epoch_start
+        samples_per_sec = rcvd_samples / epoch_dur
+        print(f"Epoch {epoch}: epoch_dur={epoch_dur:.2f}s, samples_per_sec={samples_per_sec:.2f}")
```
Comment (abhi-mosaic) [inline tests/datasets/test_streaming_remote.py:104]: Running this test with `pytest -s ...` on my Macbook Pro, with a 300Mbps connection, I see speeds like:
```
tests/datasets/test_streaming_remote.py ds_build_dur=1.45s, loader_build_dur=0.00s
Epoch 0: epoch_dur=24.28s, samples_per_sec=82.38
Epoch 1: epoch_dur=5.54s, samples_per_sec=361.02
Epoch 2: epoch_dur=5.83s, samples_per_sec=343.22
```

Our DeepLabV3 benchmark only consumes data at a rate of ~20 img/sec per A100, so if my laptop can support 80+ img/sec on epoch0 and 300+ img/sec on epoch1, I think we are in a good spot :)

**Review by abhi-mosaic — COMMENTED** [review]
@ravi-mosaicml @hanlint Went through a refactor of the ADE20k dataset and would love to hear your thoughts! After feedback I will basically do the same for ImageNet and COCO (so don't read those yet).

The 3 files to look at are:
* `composer/datasets/ade20k.py`
* `scripts/mds/ade20k.py`
* `tests/datasets/test_streaming_remote.py`

Going to start doing E2E convergence test in parallel as well!

File: composer/datasets/ade20k.py [hunk 7]
```diff
@@ -388,6 +391,155 @@ def initialize_object(self, batch_size, dataloader_hparams) -> DataSpec:
...
+    def decode_annotation(self, data: bytes) -> Any:
+        return Image.open(BytesIO(data))
+
+    def __init__(self,
+                 remote: str,
+                 local: str,
+                 split: str,
+                 shuffle: bool,
+                 base_size: int = 512,
+                 min_resize_scale: float = 0.5,
```
Comment (ravi-mosaicml) [inline composer/datasets/ade20k.py:424]: Should the dataset class be opinionated about the transformations? (Realize you likely made this change given the previous feedback on the PR!) If the answer is no, then OK to leave this as is and clean up later -- potentially once torch data pipes come out.

File: tests/datasets/test_streaming_remote.py (hunk 5 above)
Comment (ravi-mosaicml) [reply tests/datasets/test_streaming_remote.py:12]: Can you add in the integration marker in this PR? Here's how:

1. Modify `pyproject.toml` -- it'll need to be specified on line 58 (e.g. `and not integration`) and in the list on line 63

  https://github.com/mosaicml/composer/blob/23f888e07456d31a2950db9f962f4dbdf41938fb/pyproject.toml#L58-L64

2. In `meta.yaml`, modify the existing markers there to include `and not integration`

  https://github.com/mosaicml/composer/blob/97fbab157ebd8c2488f5a00a806e64639df85d19/meta.yaml#L82-L83

3. In `.ci/Jenkinsfile`, modify the markers on line 194 to also include `and not integration`

https://github.com/mosaicml/composer/blob/97fbab157ebd8c2488f5a00a806e64639df85d19/.ci/Jenkinsfile#L194

Then, the `@pytest.mark.skip` will not be needed.

**Review by ravi-mosaicml — APPROVED** [review]
LGTM -- didn't give this PR a close review since it is implementation specific, but see comments.

File: scripts/mds/ade20k.py (hunk 4 above)
Comment (Landanjs) [reply scripts/mds/ade20k.py:83]: Results on the test set require submitting to a leaderboard. This is something we should do at some point, but let me check if this should be a requirement for numbers on composer

File: composer/datasets/ade20k.py (hunk 7 above)
Comment (abhi-mosaic) [reply composer/datasets/ade20k.py:424]: I think we want to stay opinionated (and yes that was based on comments earlier in the PR :), as I think our goal is not to build an ADE20k for someone to re-use (other than basic functionality), rather, ADE20k is an example that a user will look at in order to build their own custom datasets, with custom processing.

File: composer/datasets/ade20k.py [hunk 8]
```diff
@@ -388,6 +391,136 @@ def initialize_object(self, batch_size, dataloader_hparams) -> DataSpec:
...
+        min_resize_scale (float): the minimum value the samples can be rescaled. Default: ``0.5``.
+        max_resize_scale (float): the maximum value the samples can be rescaled. Default: ``2.0``.
+        final_size (int): the final size of the image and target. Default: ``512``.
+        ignore_background (bool): if true, ignore the background class when calculating the training loss.
+            Default: ``true``.
+    """
+
+    remote: str = hp.optional('Remote directory (S3 or local filesystem) where dataset is stored',
+                              default='s3://mosaicml-internal-dataset-ade20k/mds/')
+    local: str = hp.optional('Local filesystem directory where dataset is cached during operation',
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/datasets/ade20k.py:510]: Hmm would you like the defaulting to occur in the Dataset object or the DatasetHparams ?

File: composer/datasets/ade20k.py [hunk 9]
```diff
@@ -388,6 +391,136 @@ def initialize_object(self, batch_size, dataloader_hparams) -> DataSpec:
...
+        split (str): the dataset split to use either 'train', 'val', or 'test'. Default: ``'train```.
+        base_size (int): initial size of the image and target before other augmentations. Default: ``512``.
+        min_resize_scale (float): the minimum value the samples can be rescaled. Default: ``0.5``.
+        max_resize_scale (float): the maximum value the samples can be rescaled. Default: ``2.0``.
+        final_size (int): the final size of the image and target. Default: ``512``.
+        ignore_background (bool): if true, ignore the background class when calculating the training loss.
+            Default: ``true``.
+    """
+
+    remote: str = hp.optional('Remote directory (S3 or local filesystem) where dataset is stored',
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/datasets/ade20k.py:508]: Right... would you prefer that I use a different default? Sadly I cannot remove the default as we need to use `hp.optionals` unless we clean up `DatasetHparams`. Removing these four args (which I honestly think we should do as it's a leaky abstraction) would fix it:


```
class DatasetHparams(hp.Hparams, abc.ABC, metaclass=metaclass):
    is_train: bool = hp.optional("Whether to load the training data (the default) or validation data.", default=True)
    drop_last: bool = hp.optional(textwrap.dedent("""\
        If the number of samples is not divisible by the batch size,
        whether to drop the last batch (the default) or pad the last batch with zeros."""),
                                  default=True)
    shuffle: bool = hp.optional("Whether to shuffle the dataset for each epoch. Defaults to True.", default=True)

    datadir: Optional[str] = hp.optional("The path to the data directory", default=None)
```

Alternatively I can make a whole new `StreamingDatasetHparams` that doesn't subclass `DatasetHparams` but it seems like a hack around the issue.

**Review by hanlint — APPROVED** [review]
LGTM, much cleaner, thank you!

File: composer/datasets/ade20k.py (hunk 3 above)
Comment (hanlint) [reply composer/datasets/ade20k.py]: Yeah OK to limit your refactor to streaming datasets

File: scripts/mds/ade20k.py (hunk 4 above)
Comment (abhi-mosaic) [reply scripts/mds/ade20k.py:84]: Discussed offline, will leave test set out for now.

File: tests/datasets/test_streaming_remote.py (hunk 5 above)
Comment (abhi-mosaic) [reply tests/datasets/test_streaming_remote.py]: Done!

File: composer/datasets/ade20k.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/datasets/ade20k.py]: Going to punt this to a `DatasetHparams` refactor, for now I'm just following what we did for the webdatasets.

**Comment by abhi-mosaic** [comment]
Got tests working for both `ADE20k` and `ImageNet`, I had some issues with `COCO` though so I'm going to make a JIRA issue for that and `pytest.skip()` for now. 

Adding streaming YAMLs, rebasing, and will merge once Jenkins is happy :)

**Comment by abhi-mosaic** [comment]
Some notes on ImageNet dataset performance:

On a 2xA100 node with high aggregate network bandwidth (~10Gbps):
- streaming dataloader with 8 workers/GPU can support ~3k img/sec/GPU, equivalent to 24k img/sec on an 8xA100 system, **even on epoch0**, and it gets a bit faster on epoch1+
- as expected, R50+ImageNet+channels_last on A100 **is no longer bottlenecked**, and we should have about 50% headroom for algorithms that speed up the training step
